import logging

from src.chat_service import send_system_chat
from src.pb.room_events_pb2 import RoomEvent
from src.socket_gateway import SocketGateway

logger = logging.getLogger(__name__)


async def handle_room_event(gateway: SocketGateway, event: RoomEvent):
    event_type = event.WhichOneof("event")
    match event_type:
        case "problem_solved":
            await handle_problem_solved(
                gateway,
                event.room_id,
                event.problem_solved.problem_id,
                event.problem_solved.username,
            )
        case "room_started":
            await handle_room_started(gateway, event.room_id)
        case "room_ready_failed":
            await handle_room_failed(gateway, event.room_id, event.room_ready_failed.message)
        case _:
            logger.info("ì•Œ ìˆ˜ ì—†ëŠ” ì´ë²¤íŠ¸ íƒ€ì…")


async def handle_problem_solved(
        gateway: SocketGateway,
        room_id: int,
        problem_id: str,
        username: str,
):
    await gateway.emit_problem_solved(room_id, problem_id, username)
    await send_system_chat(gateway, room_id, f"{username}ì´ {problem_id}ë¥¼ í•´ê²°í•˜ì˜€ìŠµë‹ˆë‹¤!")


async def handle_room_started(gateway: SocketGateway, room_id: int):
    await send_system_chat(gateway, room_id, "ê²Œì„ì´ ì‹œì‘ë˜ì—ˆìŠµë‹ˆë‹¤! ğŸ”¥")
    await gateway.emit_refresh(room_id)


async def handle_room_failed(gateway: SocketGateway, room_id: int, message: str):
    await send_system_chat(gateway, room_id, f"ë°© ìƒì„±ì´ ì‹¤íŒ¨í•˜ì˜€ìŠµë‹ˆë‹¤, 1ë¶„ ë’¤ ì¬ì‹œë„í•©ë‹ˆë‹¤: {message}")
    await gateway.emit_refresh(room_id)
